addons/account/models/partner.py

```diff
@@ -248,15 +230,22 @@ def _get_fiscal_position(self, partner, delivery=None):
         if manual_fiscal_position:
             return manual_fiscal_position
 
-        # First search only matching VAT positions
-        vat_valid = self._get_vat_valid(delivery, company)
-        fp = self._get_fpos_by_region(delivery.country_id.id, delivery.state_id.id, delivery.zip, vat_valid)
-
-        # Then if VAT required found no match, try positions that do not require it
-        if not fp and vat_valid:
-            fp = self._get_fpos_by_region(delivery.country_id.id, delivery.state_id.id, delivery.zip, False)
-
-        return fp or self.env['account.fiscal.position']
+        # Search for a an auto applied fiscal position matching the partner
+        ranking_subfunctions = self._get_fpos_ranking_functions(delivery)
+        def ranking_function(fpos):
+            return tuple(rank[1](fpos) for rank in ranking_subfunctions)
+
+        return max(
+            filter(lambda x: all(x[1]), [
+                (fpos, ranking_function(fpos))
+                for fpos in self.search([
+                    *self._check_company_domain(self.env.company),
+                    ('auto_apply', '=', True),
+                ])
+            ]),
+            key=lambda x: x[1],
+            default=(self.env['account.fiscal.position'], False)
+        )[0]
```

Comment:
Several stuff:
- if there isn't any fiscal position satisfying every criteria, it won't be selected (because of the `all(critera_tuple)`)
At first sight, it's nice because it means you will get the perfect one but what if there isn't any that correspond but one is nearly the good one? Do we want to not set any because it's not close enough?
- This looks cleaver but is very hard to read. It would be better to use intermediary variables like `all_auto_apply_fpos` for the search, `ranked_fpos` for the list comp, `eligible_fpos` for the filter.
- nitpick: list comp could be a gen comp, no need to materialize the list